packages/ckeditor5-engine/tests/model/schema.js

```diff
@@ -768,59 +790,69 @@ describe( 'Schema', () => {
+			schema.register( 'paragraph', { allowIn: '$root' } );
+			schema.register( 'blockQuote', { allowIn: '$root' } );
+			schema.register( 'foo' );
 		} );
 
-		it( 'adds a high-priority listener', () => {
+		it( 'supports generic and specific checks', () => {
+			const spyGeneric = sinon.spy();
+			const spySpecificP = sinon.spy();
+			const spySpecificBar = sinon.spy();
+
+			schema.addChildCheck( spyGeneric );
+			schema.addChildCheck( spySpecificP, 'paragraph' );
+			schema.addChildCheck( spySpecificBar, 'bar' );
+
+			schema.checkChild( [ '$root' ], 'paragraph' );
+
+			expect( spyGeneric.calledOnce ).to.be.true;
+			expect( spySpecificP.calledOnce ).to.be.true;
+			expect( spySpecificBar.called ).to.be.false;
+		} );
+
+		it( 'proper checks order', () => {
 			const order = [];
 
-			schema.on( 'checkChild', () => {
-				order.push( 'checkChild:high-before' );
-			}, { priority: 'high' } );
+			schema.addChildCheck( () => {
+				order.push( 'addChildCheckSpecific' );
+			}, 'paragraph' );
 
 			schema.addChildCheck( () => {
-				order.push( 'addChildCheck' );
+				order.push( 'addChildCheckGeneric' );
 			} );
 
 			schema.on( 'checkChild', () => {
-				order.push( 'checkChild:high-after' );
+				order.push( 'checkChild:high' );
 			}, { priority: 'high' } );
 
 			schema.checkChild( root1, r1p1 );
 
-			expect( order.join() ).to.equal( 'checkChild:high-before,addChildCheck,checkChild:high-after' );
+			expect( order.join() ).to.equal( 'checkChild:high,addChildCheckGeneric,addChildCheckSpecific' );
 		} );
 
-		it( 'stops the event and overrides the return value when callback returned true', () => {
-			schema.register( '$text' );
+		it( 'overrides the declarative check - force allow', () => {
+			expect( schema.checkChild( [ '$root' ], 'foo' ) ).to.be.false;
 
-			expect( schema.checkChild( root1, '$text' ) ).to.be.false;
+			schema.addChildCheck( () => true, 'foo' );
 
-			schema.addChildCheck( () => {
-				return true;
-			} );
+			expect( schema.checkChild( [ '$root' ], 'foo' ) ).to.be.true;
+		} );
 
-			schema.on( 'checkChild', () => {
-				throw new Error( 'the event should be stopped' );
-			}, { priority: 'high' } );
+		it( 'overrides the declarative check - force disallow', () => {
+			expect( schema.checkChild( [ '$root' ], 'paragraph' ) ).to.be.true;
 
-			expect( schema.checkChild( root1, '$text' ) ).to.be.true;
-		} );
+			schema.addChildCheck( () => false, 'paragraph' );
 
-		it( 'stops the event and overrides the return value when callback returned false', () => {
-			expect( schema.checkChild( root1, r1p1 ) ).to.be.true;
+			expect( schema.checkChild( [ '$root' ], 'paragraph' ) ).to.be.false;
+		} );
 
-			schema.addChildCheck( () => {
-				return false;
-			} );
+		it( 'uses declarative check when undefined is returned', () => {
+			expect( schema.checkChild( [ '$root' ], 'paragraph' ) ).to.be.true;
 
-			schema.on( 'checkChild', () => {
-				throw new Error( 'the event should be stopped' );
-			}, { priority: 'high' } );
```

Comment:
Shouldn't this be marked as a breaking change? Same as for the attribute check?
What if someone provided a child check callback or a custom listener who relied on the event being stopped before?

Reply:
I tried to analyze this and you are right, and I added a minor breaking change entry.

However, if I get it correctly, the scope for possible problem is quite narrow here.

If somebody was using only direct event listeners (i.e. `schema.on( 'checkChild' )`) then it seems to me that nothing changed for them. They were probably adding these events on `high` priority and were stopping them (otherwise, these callback would not make much sense, I think).

If somebody was using only callbacks added through `addChildCheck()` or `addAttributeCheck()` then nothing changes here as well -- they are still working in "first wins" fashion, even though the event is not stopped.

So if I get it right, the only problem may happen if somebody was defining rules both using the event listener and using the methods, because they might have first define `addChildCheck( callback )` and then add listener with `high` priority, and in such case, if `callback` returned a value, the listener would not fire. But now, not only the listener will fire, it will fire before the callback.

Reply:
Let me know what you think about the entry. I first described it for "child checks" and then at the end added that it is true also for "attribute checks". When I was describing both at once, it became almost unreadable.